File: src/Butil/Bit.Butil/Scripts/digitalGoods.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    const DEFAULT_PROVIDER = 'https://play.google.com/billing';

    // One connection per provider, kept for the life of the page: getDigitalGoodsService() is a
    // connection to the store, not a lookup, and reconnecting on every call would put a round trip
    // in front of each one. What is kept is the pending promise, not the resolved service, so
    // concurrent first calls - listing purchases and fetching the catalogue together on launch - share
    // one connection attempt instead of each opening a connection and orphaning all but the last.
    const _services: { [provider: string]: Promise<any> } = {};

    butil.digitalGoods = {
        isSupported() { return typeof (window as any).getDigitalGoodsService === 'function'; },
        connect,
        getDetails,
        listPurchases,
        listPurchaseHistory,
        consume
    };

    function service(provider: string): Promise<any> {
        const W = window as any;
        if (typeof W.getDigitalGoodsService !== 'function') return Promise.resolve(null);

        const key = provider || DEFAULT_PROVIDER;
        return _services[key] || (_services[key] = openService(W, key));
    }

    async function openService(W: any, key: string) {
        try {
            const connected = await W.getDigitalGoodsService(key);
            if (connected) return connected;
        } catch {
            // Not installed from this store - the ordinary answer in a browser tab.
        }

        delete _services[key];
        return null;
    }

    // A store call that rejects may mean the connection behind it is gone (the store app restarted),
    // and a cached dead connection would answer every later call with nothing until a reload. Dropping
    // it makes the next call reconnect.
    function forget(provider: string) {
        delete _services[provider || DEFAULT_PROVIDER];
    }

    function toAmount(amount: any) {
        return amount ? { currency: amount.currency ?? '', value: amount.value ?? '' } : null;
    }

    function toPurchase(purchase: any) {
        return { itemId: purchase.itemId ?? '', purchaseToken: purchase.purchaseToken ?? '' };
    }

    async function connect(provider: string) {
        return !!(await service(provider));
    }

    async function getDetails(itemIds: string[], provider: string) {
        const store = await service(provider);
        if (!store?.getDetails || !itemIds?.length) return [];

        try {
            const items = await store.getDetails(itemIds);
            return (items || []).map((item: any) => ({
                itemId: item.itemId ?? '',
                title: item.title ?? '',
                description: item.description ?? null,
                price: toAmount(item.price),
                type: item.type ?? null,
                // iconURLs in the specification; Chrome shipped iconUrls first and still answers to it.
                iconUrls: item.iconURLs ?? item.iconUrls ?? [],
                subscriptionPeriod: item.subscriptionPeriod ?? null,
                freeTrialPeriod: item.freeTrialPeriod ?? null,
                introductoryPrice: toAmount(item.introductoryPrice),
                introductoryPricePeriod: item.introductoryPricePeriod ?? null,
                introductoryPriceCycles: item.introductoryPriceCycles ?? null
            }));
        } catch {
            forget(provider);
            return [];
        }
    }

    async function listPurchases(provider: string) {
        const store = await service(provider);
        if (!store?.listPurchases) return [];

        try { return ((await store.listPurchases()) || []).map(toPurchase); }
        catch { forget(provider); return []; }
    }

    async function listPurchaseHistory(provider: string) {
        const store = await service(provider);
        if (!store?.listPurchaseHistory) return [];

        try { return ((await store.listPurchaseHistory()) || []).map(toPurchase); }
        catch { forget(provider); return []; }
    }

    async function consume(purchaseToken: string, provider: string) {
        const store = await service(provider);
        if (!store?.consume || !purchaseToken) return false;

        try { await store.consume(purchaseToken); return true; }
        catch { forget(provider); return false; }
    }
}(BitButil));

File: src/Butil/Bit.Butil/Publics/DigitalGoods.cs
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/Digital_Goods_API">Digital Goods API</see>
/// (<c>window.getDigitalGoodsService()</c>): reads an app store's catalogue and the user's
/// entitlements from inside an installed PWA.
/// </summary>
/// <remarks>
/// This half only queries - it lists items, purchases and history, and consumes a purchase. Buying
/// is a <see cref="PaymentRequest"/> with the store's payment method identifier
/// (<c>"https://play.google.com/billing"</c>) and the item id in its data, and the purchase shows up
/// here afterwards.
/// <br/>
/// Chromium only, and only in an app installed from the store it is asking - a browser tab has no
/// store to connect to, so <see cref="Connect"/> returns false there. Every entitlement decision
/// belongs on your server, verified against the store's API with a
/// <see cref="DigitalGoodsPurchase.PurchaseToken"/>; a client that says it paid is a claim, not a
/// receipt.
/// </remarks>
[ButilService(typeof(DigitalGoods))]
public class DigitalGoods(IJSRuntime js)
{
    /// <summary>
    /// The payment method identifier for Google Play billing - the service provider these calls
    /// default to, and the method identifier a <see cref="PaymentRequest"/> buys through.
    /// </summary>
    public const string GooglePlayBilling = "https://play.google.com/billing";

    /// <summary>
    /// True when the runtime exposes <c>window.getDigitalGoodsService</c>.
    /// <br/>
    /// <see href="https://developer.mozilla.org/en-US/docs/Web/API/Window/getDigitalGoodsService">Window.getDigitalGoodsService()</see>
    /// </summary>
    /// <remarks>
    /// Being exposed is not the same as being connectable: an uninstalled app has the function and
    /// no store behind it. <see cref="Connect"/> is the check that answers for real.
    /// <br/>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.digitalGoods.isSupported");

    /// <summary>
    /// Connects to a store's billing service, returning false when there is none - the usual answer
    /// in a browser tab. The connection is cached, so the other members can be called straight away
    /// and this is only worth calling to decide whether to show a store UI at all.
    /// <br/>
    /// <see href="https://developer.mozilla.org/en-US/docs/Web/API/Window/getDigitalGoodsService">Window.getDigitalGoodsService()</see>
    /// </summary>
    /// <param name="serviceProvider">The store's payment method identifier. Defaults to <see cref="GooglePlayBilling"/>.</param>
    public ValueTask<bool> Connect(string serviceProvider = GooglePlayBilling)
        => js.Invoke<bool>("BitButil.digitalGoods.connect", serviceProvider);

    /// <summary>
    /// Looks up the store's own title, description and price for each item id, skipping ids the
    /// store does not know. Empty when there is no store to ask.
    /// <br/>
    /// <see href="https://developer.mozilla.org/en-US/docs/Web/API/DigitalGoodsService/getDetails">DigitalGoodsService.getDetails()</see>
    /// </summary>
    /// <param name="itemIds">The item ids configured in the store's console.</param>
    /// <param name="serviceProvider">The store's payment method identifier. Defaults to <see cref="GooglePlayBilling"/>.</param>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(DigitalGoodsItem))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(PaymentCurrencyAmount))]
    public ValueTask<DigitalGoodsItem[]> GetDetails(string[] itemIds, string serviceProvider = GooglePlayBilling)
        => js.Invoke<DigitalGoodsItem[]>("BitButil.digitalGoods.getDetails", itemIds, serviceProvider);

    /// <summary>
    /// The purchases that are currently active - unconsumed one-offs and live subscriptions. This is
    /// what a client restores entitlements from after a reinstall, and what it re-checks on launch.
    /// <br/>
    /// <see href="https://developer.mozilla.org/en-US/docs/Web/API/DigitalGoodsService/listPurchases">DigitalGoodsService.listPurchases()</see>
    /// </summary>
    /// <param name="serviceProvider">The store's payment method identifier. Defaults to <see cref="GooglePlayBilling"/>.</param>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(DigitalGoodsPurchase))]
    public ValueTask<DigitalGoodsPurchase[]> ListPurchases(string serviceProvider = GooglePlayBilling)
        => js.Invoke<DigitalGoodsPurchase[]>("BitButil.digitalGoods.listPurchases", serviceProvider);

    /// <summary>
    /// The most recent purchase per item, including ones already consumed or expired - the history
    /// behind <see cref="ListPurchases"/> rather than the current state.
    /// <br/>
    /// <see href="https://developer.mozilla.org/en-US/docs/Web/API/DigitalGoodsService/listPurchaseHistory">DigitalGoodsService.listPurchaseHistory()</see>
    /// </summary>
    /// <param name="serviceProvider">The store's payment method identifier. Defaults to <see cref="GooglePlayBilling"/>.</param>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(DigitalGoodsPurchase))]
    public ValueTask<DigitalGoodsPurchase[]> ListPurchaseHistory(string serviceProvider = GooglePlayBilling)
        => js.Invoke<DigitalGoodsPurchase[]>("BitButil.digitalGoods.listPurchaseHistory", serviceProvider);

    /// <summary>
    /// Marks a purchase as used up, so the item can be bought again - the coin pack the user spent.
    /// Returns false when the store refused or there was no store.
    /// <br/>
    /// <see href="https://developer.mozilla.org/en-US/docs/Web/API/DigitalGoodsService/consume">DigitalGoodsService.consume()</see>
    /// </summary>
    /// <param name="purchaseToken">The <see cref="DigitalGoodsPurchase.PurchaseToken"/> to consume.</param>
    /// <param name="serviceProvider">The store's payment method identifier. Defaults to <see cref="GooglePlayBilling"/>.</param>
    /// <remarks>
    /// Consume only after your server has recorded what the purchase bought. Consuming first and
    /// crediting second loses the purchase entirely if anything in between fails.
    /// </remarks>
    public ValueTask<bool> Consume(string purchaseToken, string serviceProvider = GooglePlayBilling)
        => js.Invoke<bool>("BitButil.digitalGoods.consume", purchaseToken, serviceProvider);
}
